status/status.go

```diff
@@ -97,32 +97,29 @@ func FromError(err error) (s *Status, ok bool) {
 	if err == nil {
 		return nil, true
 	}
+	defer func() {
+		if r := recover(); r != nil {
+			// type assertion passes but GRPCStatus method is still nil
+			s = New(codes.Unknown, err.Error())
+			ok = false
+		}
+	}()
 	type grpcstatus interface{ GRPCStatus() *Status }
-	if gs, ok := err.(grpcstatus); ok {
-		grpcStatus := gs.GRPCStatus()
-		if grpcStatus == nil {
-			// Error has status nil, which maps to codes.OK. There
-			// is no sensible behavior for this, so we turn it into
-			// an error with codes.Unknown and discard the existing
-			// status.
-			return New(codes.Unknown, err.Error()), false
+	if gs, ok := err.(grpcstatus); ok && gs != nil {
```

Comment:
Isn't the `gs != nil` unnecessary here now, then?

I'm not sure this actually catches anything.  So maybe we can keep the rest of the code as is was before and just add the defer/recover?